pypose/lietensor/operation.py

```diff
@@ -328,4 +457,291 @@ def backward(ctx, grad_output):
+
+
+class SE3_mul(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, X, Y):
+        ctx.save_for_backward(X)
+        t = X[..., :3] + SO3_Act.apply(X[..., 3:], Y[..., :3])
+        q = SO3_mul.apply(X[..., 3:], Y[..., 3:])
+        return torch.cat((t, q), -1)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        X = ctx.saved_tensors[0]
+        zero = torch.zeros(X.shape[:-1]+(1,), device=X.device, dtype=X.dtype)
+        X_grad = torch.cat((grad_output[..., :-1], zero), dim = -1)
+        dZxX = torch.matmul(grad_output[..., :-1].unsqueeze(-2), SE3_Adj(X)).squeeze(-2) 
+        Y_grad = torch.cat((dZxX, zero), dim = -1)
+        return X_grad, Y_grad        
+
+
+class RxSO3_mul(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, X, Y):
+        ctx.save_for_backward(X)
+        q = SO3_mul.apply(X[..., :4], Y[..., :4])
+        s = X[..., 4:] * Y[..., 4:]
+        return torch.cat((q, s), -1)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        X = ctx.saved_tensors[0]
+        zero = torch.zeros(X.shape[:-1]+(1,), device=X.device, dtype=X.dtype)
+        X_grad = torch.cat((grad_output[..., :-1], zero), dim = -1)
+        dZxX = torch.matmul(grad_output[..., :-1].unsqueeze(-2), RxSO3_Adj(X)).squeeze(-2) 
+        Y_grad = torch.cat((dZxX, zero), dim = -1)
+        return X_grad, Y_grad    
+
+
+class Sim3_mul(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, X, Y):
+        ctx.save_for_backward(X)
+        t = X[..., :3] + RxSO3_Act.apply(X[..., 3:], Y[..., :3])
+        q = RxSO3_mul.apply(X[..., 3:], Y[..., 3:])
+        return torch.cat((t, q), -1)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        X = ctx.saved_tensors[0]
+        zero = torch.zeros(X.shape[:-1]+(1,), device=X.device, dtype=X.dtype)
+        X_grad = torch.cat((grad_output[..., :-1], zero), dim = -1)
+        dZxX = torch.matmul(grad_output[..., :-1].unsqueeze(-2), Sim3_Adj(X)).squeeze(-2) 
+        Y_grad = torch.cat((dZxX, zero), dim = -1)
+        return X_grad, Y_grad   
+
+
+def broadcast_inputs(x, y):
+    """ Automatic broadcasting of missing dimensions """
+    if y is None:
+        xs, xd = x.shape[:-1], x.shape[-1]
+        return (x.reshape(-1, xd).contiguous(), ), x.shape[:-1]
+    out_shape = torch.broadcast_shapes(x.shape[:-1], y.shape[:-1])
+    shape = out_shape if out_shape != torch.Size([]) else (1,)
+    x = x.expand(shape+(x.shape[-1],)).reshape(-1,x.shape[-1]).contiguous()
+    y = y.expand(shape+(y.shape[-1],)).reshape(-1,y.shape[-1]).contiguous()
+    return (x, y), tuple(out_shape)
+
+
+def lietensor_act(lid, x, p):
+    x = x.tensor() if hasattr(x, 'ltype') else x
+    input, out_shape = broadcast_inputs(x, p)
+    if lid == 1:
+        out = SO3_Act.apply(*input)
+    elif lid == 2:
+        out = RxSO3_Act.apply(*input)
+    elif lid == 3:
+        out = SE3_Act.apply(*input)
+    elif lid == 4:
+        out = Sim3_Act.apply(*input)
```

Comment:
Shall we distribute the `lid`-defined operations to each their corresponding classes, similar to what we did for `Exp` and `Log`, then `lid` won't be needed? We can remove the `lid` definition when all operations are defined within Python.

The same questions apply to `lietensor_act4` and `lietensor_mul` also.